### DISCUSSION TITLE
What am I doing wrong? :) Slow writes to db.

### DISCUSSION THREAD

**Discussion by mdrgeniebelt** [prisma/prisma#8729] [Q&A]
Hi, I have a question regarding performance of writes to db using prisma client. 

So I created very simple app in nodejs to test performance of writing less than 10 elements to db - postgres, and I noticed that it takes around 70ms to write 11 rows into database. I added some logging to see what is taking so long and it seems that queries are very fast so what is impacting the overall time it takes to finish the transaction?

Here is the sample code:

Schema
```prisma
datasource db {
  url      = ""
  provider = "postgresql"
}

generator client {
  provider = "prisma-client-js"
}

model Task {
  id        String    @id @default(uuid())
  createdAt DateTime  @default(now())
  name      String?   @default("")
  startDate DateTime?
  endDate   DateTime?
  Account   Account?   @relation(fields: [accountId], references: [id])
  accountId String?
}

model Account {
  id        String   @id @default(uuid())
  createdAt DateTime @default(now())
  name      String?  @default("")
  email     String
  Task      Task[]
}
```

And simple ts snippet:
```js
import { Prisma, PrismaClient } from "@prisma/client";
import { v4 as uuid } from "uuid";
import perfy from "perfy";

const prisma = new PrismaClient({
  errorFormat: "minimal",
  log: [
    {
      emit: 'event',
      level: 'query',
    }
  ]
});

prisma.$on('query', (e) => {
  console.log('Query: ' + e.query)
  console.log('Duration: ' + e.duration + 'ms')
})

const account: Prisma.AccountCreateInput = {
  email: "test@test.com",
  createdAt: new Date(),
  name: "test",
  id: uuid(),
};

const createManyTasks = (count: number): Prisma.TaskCreateInput[] => {
  const tasks: Prisma.TaskCreateInput[] = [];
  for (let i = 0; i < count; i++) {
    tasks.push({
      createdAt: new Date(),
      name: "task" + i,
      startDate: new Date(),
      endDate: new Date(),
      id: uuid(),
    });
  }
  return tasks;
}

const tasks = createManyTasks(10);

perfy.start("transaction");
prisma.account.create({
  data: {
    id: account.id,
    email: account.email,
    name: account.name,
    createdAt: account.createdAt,
    Task: {
      createMany: {
        data: tasks
      }
    }
  },
})
.then(() => {
  let res = perfy.end("transaction");
  console.log(`ended in ${res.fullMilliseconds} ms`)
})
```

It prints something like:

```
Query: BEGIN
Duration: 1ms
Query: INSERT INTO "public"."Account" ("id","createdAt","name","email") VALUES ($1,$2,$3,$4) RETURNING "public"."Account"."id"
Duration: 2ms
Query: INSERT INTO "public"."Task" ("endDate","id","name","accountId","startDate","createdAt") VALUES ($1,$2,$3,$4,$5,$6), ($7,$8,$9,$10,$11,$12), ($13,$14,$15,$16,$17,$18), ($19,$20,$21,$22,$23,$24), ($25,$26,$27,$28,$29,$30), ($31,$32,$33,$34,$35,$36), ($37,$38,$39,$40,$41,$42), ($43,$44,$45,$46,$47,$48), ($49,$50,$51,$52,$53,$54), ($55,$56,$57,$58,$59,$60)
Duration: 2ms
Query: SELECT "public"."Account"."id", "public"."Account"."createdAt", "public"."Account"."name", "public"."Account"."email" FROM "public"."Account" WHERE "public"."Account"."id" = $1 LIMIT $2 OFFSET $3
Duration: 1ms
Query: COMMIT
Duration: 1ms
ended in 75.115 ms
```

I'm not a nodejs expert but might it be because of nodejs and it being single-threaded? Any help here would be appreciated.

**Comment by ryands17** [comment]
@mdrgeniebelt 👋 

One thing you could try is enabling `nApi` that should make it much faster. Add this to the `generator` block:

```prisma
generator client {
  provider = "prisma-client-js"
  previewFeatures  = ["nApi"]
}
```

Then run `prisma generate` and test the above code. Should be faster then.

**Reply by mdrgeniebelt** [reply]
Hey @ryands17 thanks for suggestions, I did test it but it's still around 40ms

```
Query: BEGIN
Duration: 1ms
Query: INSERT INTO "public"."Account" ("id","createdAt","name","email") VALUES ($1,$2,$3,$4) RETURNING "public"."Account"."id"
Duration: 2ms
Query: INSERT INTO "public"."Task" ("createdAt","endDate","startDate","id","accountId","name") VALUES ($1,$2,$3,$4,$5,$6), ($7,$8,$9,$10,$11,$12), ($13,$14,$15,$16,$17,$18), ($19,$20,$21,$22,$23,$24), ($25,$26,$27,$28,$29,$30), ($31,$32,$33,$34,$35,$36), ($37,$38,$39,$40,$41,$42), ($43,$44,$45,$46,$47,$48), ($49,$50,$51,$52,$53,$54), ($55,$56,$57,$58,$59,$60)
Duration: 2ms
Query: SELECT "public"."Account"."id", "public"."Account"."createdAt", "public"."Account"."name", "public"."Account"."email" FROM "public"."Account" WHERE "public"."Account"."id" = $1 LIMIT $2 OFFSET $3
Duration: 1ms
Query: COMMIT
Duration: 1ms
ended in 42.299877 ms
```

First one took over 300ms after I ran `prisma generate` then each consequent run took around 40-45ms

**Reply by ryands17** [reply]
40 ms should be fine in this case. I will let @janpio confirm though.

**Reply by janpio** [reply]
Hard to say what is taking this time. Generally of course you want the time to be closer to the real time needed by the queries itself. Sources for errors might be the additional logging, the reported numbers for the queries and so on - so we would need to dig far deeper to be able to understand this.

If this is a real problem, please open a bug report and someone can take a look.

**Reply by mdrgeniebelt** [reply]
I turned off logging and it still takes around 40ms. Like you wrote, it's hard to say what might be affecting that.

Thanks for the responses anyways :)

**Reply by janpio** [reply]
Meh, shame.

Turns this into an issue, what you posted above should probably already be enough for us to reproduce and if we get equally funny numbers, to look into this deeper.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
